Wide cell boundary conditions in ECH & DCH + soft-wrap reset correctness

qwerasd205 opened the pull request:
Adds tests for boundary conditions involving wide chars for ECH & DCH and fixes them with a helper function `Screen.splitCellBoundary`.

Also introduces `Screen.cursorResetWrap` which resets the wrap state for the cursor row (and the cursor's pending wrap) in a correct way, accounting for spacer heads and the previous and next row's wrap flags as well. A few functions in `Terminal` previously just did `self.screen.cursor_row.wrap = false` and called it a day which can result in orphaned spacer heads and improper `wrap` or `wrap_continuation` flags on the prev and next rows.

An important `TODO` I left is in `eraseChars`, technically the boundary condition handling is still insufficient there, since it's not accounting for protected cells creating additional boundaries, or eliminating the two boundaries that are actually being handled at all.

qwerasd205 committed:
test(Terminal): test DCH wide char boundary conditions

qwerasd205 committed:
test(Terminal): test ECH wide char boundary conditions

qwerasd205 committed:
Terminal: fix ECH & DCH wide char boundary cond. behavior

mitchellh reviewed (changes requested):
Thanks, the tests look good, some feedback on the other parts.

qwerasd205 committed:
perf(terminal): specialize `splitCellBoundary` to cursor row

+ do some abstraction leakage in `cursorResetWrap`, since they're both
used in hot functions for TUI stuff so performance is important.

mitchellh committed:
terminal: add test for wide character on right margin boundary

qwerasd205 committed:
terminal: cursorResetWrap should not reset wrap_continuation

qwerasd205 committed:
comment
